Objects/frameobject.c

```diff
@@ -16,6 +16,622 @@
+            return -1;
+        }
+        frame->f_extra_locals = extra;
+    }
+
+    assert(PyDict_Check(extra));
+
+    if (PyDict_SetItem(extra, key, value) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static PyObject*
+framelocalsproxy_contains(PyObject *self, PyObject *key)
+{
+    PyObject* keys = framelocalsproxy_keys(self, NULL);
+
+    if (keys == NULL) {
+        return NULL;
+    }
+
+    int result = PySequence_Contains(keys, key);
+
+    Py_DECREF(keys);
+
+    if (result < 0) {
+        return NULL;
+    }
+
+    return PyBool_FromLong(result);
+}
+
+static PyObject*
+framelocalsproxy_update(PyObject *self, PyObject *other)
+{
+    if (framelocalsproxy_merge(self, other) < 0) {
+        PyErr_SetString(PyExc_TypeError, "update() argument must be dict or another FrameLocalsProxy");
+        return NULL;
+    }
+
+    Py_RETURN_NONE;
+}
+
+static PyObject*
+framelocalsproxy_get(PyObject* self, PyObject *const *args, Py_ssize_t nargs)
+{
+    if (nargs < 1 || nargs > 2) {
+        PyErr_SetString(PyExc_TypeError, "get expected 1 or 2 arguments");
+        return NULL;
+    }
+
+    PyObject *key = args[0];
+    PyObject *default_value = Py_None;
+
+    if (nargs == 2) {
+        default_value = args[1];
+    }
+
+    if (!PyUnicode_CheckExact(key)) {
+        PyErr_SetString(PyExc_TypeError, "key must be a string");
+        return NULL;
+    }
+
+    PyObject *result = framelocalsproxy_getitem(self, args[0]);
+
+    if (result == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+            PyErr_Clear();
+            return Py_XNewRef(default_value);
+        }
+        return NULL;
+    }
+
+    return result;
+}
+
+static PyObject*
+framelocalsproxy_setdefault(PyObject* self, PyObject *const *args, Py_ssize_t nargs)
+{
+    if (nargs < 1 || nargs > 2) {
+        PyErr_SetString(PyExc_TypeError, "setdefault expected 1 or 2 arguments");
+        return NULL;
+    }
+
+    PyObject *key = args[0];
+    PyObject *default_value = Py_None;
+
+    if (nargs == 2) {
+        default_value = args[1];
+    }
+
+    if (!PyUnicode_CheckExact(key)) {
+        PyErr_SetString(PyExc_TypeError, "key must be a string");
+        return NULL;
+    }
+
+    PyObject *result = framelocalsproxy_getitem(self, key);
+
+    if (result == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+            PyErr_Clear();
+            if (framelocalsproxy_setitem(self, key, default_value) < 0) {
+                return NULL;
+            }
+            return Py_XNewRef(default_value);
+        }
+        return NULL;
+    }
+
+    return result;
+}
+
+static PyObject*
+framelocalsproxy_reversed(PyObject *self, PyObject *__unused)
+{
+    PyObject *result = framelocalsproxy_keys(self, NULL);
+    if (PyList_Reverse(result) < 0) {
+        Py_DECREF(result);
+        return NULL;
+    }
+    return result;
+}
+
+static PyObject*
+framelocalsproxy_copy_unsupported(PyObject *self, PyObject *__unused)
```

Comment:
Because I can't think of the difference between `proxy_copy = copy(proxy)` and `proxy_copy = proxy`. It's confusing because even after the copy, changes to the copied version still impacts the original one. The reason to implement this is to avoid `copy` to use pickle on it and generate an error about not being able to pickle it. Still for user experience.

Reply:
It occurred to me that there's a compelling reason to implement `proxy.copy()` and `copy(proxy)` as equivalents to `dict(proxy)`: improving backwards compatibility with existing frame processing code that expects `frame.f_locals` to be a regular dict with some implementation defined implicit update rules.

In previous versions, you could use any of those dict copying operations to break the link to the dict cached on the frame object, so we may as well support all three of them in 3.13+ as a way to switch from a proxy instance to a dict snapshot.

Reply:
> It occurred to me that there's a compelling reason to implement `proxy.copy()` and `copy(proxy)` as equivalents to `dict(proxy)`: improving backwards compatibility with existing frame processing code that expects `frame.f_locals` on an optimised frame to be a regular dict with some implementation defined implicit update rules.
> 
> In previous versions, you could use any of those dict copying operations to break the link to the dict cached on the frame object, so we should support all three of them in 3.13+ as a way to switch from a proxy instance to a dict snapshot.

I'm not so sure. We know that code that works on 3.12 won't *necessarily* work on 3.13, so we have to accept breaking some code. If we just support `dict(f.f_locals)`, that allows a copy to be made on all versions using that syntax. Making `f.f_locals.copy()` or `copy(f.f_locals)` also return a dict will forever block us from implementing those operations for proxies in a more useful way.

Reply:
I kind of agree with Guido here. We are proposing a PEP for `f_locals` because it's a big change, one that might not be backwards compatible. The PEP is approved because the new way is much more explainable and much less confusing. Having a "copy" that is not the type of the original feels wrong to me and it could confuse users as well - is it obvious too all people that a "copied" proxy is not linked to the original frame locals? I have some doubts. `dict(f.f_locals)` however, has a clearer semantic that this would be a pure snapshot.

Also, `frame.f_locals` is mostly used by tool developers who should be more familiar with Python itself than most people. I think it's okay to ask them to adapt to this new feature even if in some aspects it's not fully backwards compatible.